Add scalar function DESTRUCTURE_TDIGEST

sdimmick opened the pull request:
* Add `sum` to TDigest and update serialization version to handle new field
* Add new scalar function on TDigest type to expose the "raw" TDigest data
structure. The prototype of the new function is:
```
DESTRUCTURE_TDIGEST(column: TDIGEST) ->

ROW(
  centroid_means: ARRAY[DOUBLE],
  centroid_weights: ARRAY[INTEGER],
  compression: DOUBLE,
  min: DOUBLE,
  max: DOUBLE,
  sum: DOUBLE,
  count: INTEGER
)
```

Test plan

`mvn -Dtest="TestTDigest,TestTDigestFunctions,TestMergeTDigestFunction" test`

```
== RELEASE NOTES ==

General Changes
* Add new scalar function on TDigest type to expose the "raw" TDigest data
structure. The prototype of the new function is:

DESTRUCTURE_TDIGEST(column: TDIGEST) ->

ROW(
  centroid_means: ARRAY[DOUBLE],
  centroid_weights: ARRAY[INTEGER],
  compression: DOUBLE,
  min: DOUBLE,
  max: DOUBLE,
  sum: DOUBLE,
  count: INTEGER
)
```

tdcmeehan commented:
How about adding a `compression` field, which is a `DOUBLE`, so that this can return the entire internal representation of the T-Digest?  And then renaming this to something like `DESTRUCTURE_TDIGEST`.

sdimmick commented:
> How about adding a `compression` field, which is a `DOUBLE`, so that this can return the entire internal representation of the T-Digest? And then renaming this to something like `DESTRUCTURE_TDIGEST`.

Great idea. Addressed in 029870cdfa08fa8e57d065087ba2863ff0736cee. I've also updated the PR title / description.

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/TDigestFunctions.java:
```diff
@@ -86,4 +104,40 @@ public static Slice scaleTDigestDouble(@SqlType("tdigest(double)") Slice input,
+    {
+        TDigest tDigest = createTDigest(input);
+
+        PageBuilder pageBuilder = new PageBuilder(ImmutableList.of(TDIGEST_CENTROIDS_ROW_TYPE));
```
You don't really need to create a `PageBuilder` here. Please directly create a `RowBlockBuilder`. You can use `RowType.createBlockBuilder()`

sdimmick replied:
Thanks for the suggestion. Addressed in db91cce971fce65dd097de1a089ec92f49f9be8a

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/TDigestFunctions.java:
```diff
@@ -86,4 +104,40 @@ public static Slice scaleTDigestDouble(@SqlType("tdigest(double)") Slice input,
         return digest.serialize();
     }
+
+    @ScalarFunction(value = "destructure_tdigest", visibility = EXPERIMENTAL)
```
Can we have a better name. @tdcmeehan any suggestions?

tdcmeehan replied:
I can't think of anything better.  To me this is destructuring the internal representation into a struct.

Maybe we can add this as an operator instead, and represent this as a cast?

rongrong replied:
@kaikalur Would `CAST` fit the bill here? If you implement this as `cast` operator we can potentially allow user to pick and choose the fields. Though I'm not sure whether the behavior would be consistent with the spec?

kaikalur replied:
I'm not an expert on what a user would want here. If we think they would want individual fields, maybe we abstract from the fact that this is a row and give them a function for each of the fields:

TDIGEST_EXTRACT_SUM(tdigest): DOUBLE

etc.

rongrong reviewed (commented):
We should also add documentation for add this function. @tdcmeehan have we documented all tdigest functions?

tdcmeehan commented:
> We should also add documentation for add this function. @tdcmeehan have we documented all tdigest functions?

No.  We should add formal documentation.

sdimmick commented:
It seems like the CI build is failing due to unrelated tests failing. I tried merging master, and that build failed on different tests. Any advice?

sdimmick commented:
@tdcmeehan @rongrong is there anything I can do to help resolve the naming discussion? I'm happy to offer suggestions of my own, or to go with any of yours!

kaikalur reviewed (commented):
Looks good. I would like to see an actual query thoough

kaikalur commented:
Discussed offline about function vs struct so for now this is fine.

kaikalur reviewed (approved):
+1

rongrong reviewed (approved):
LGTM. @tdcmeehan  Please merge if you think this is good too!

prithvip commented:
@sdimmick Can we add release notes for this change? Thanks! cc @tdcmeehan

sdimmick commented:
> @sdimmick Can we add release notes for this change? Thanks! cc @tdcmeehan

@prithvip I added release notes to the PR description per the default template. Did I mess that up?  Or is there another way I should submit them?
